### ISSUE TITLE
setup.py install_requires missing numpy

### ISSUE THREAD

**Issue by Speedy1991** [inducer/pycuda#147]
setup.py is missing numpy, installing with pip leads to:
`ImportError: No module named 'numpy'`
running pip with python3.6

**Comment by inducer** [comment]
It actually needs `numpy` before it gets to the setuptools bit, so adding it to `install_requires` wouldn't help. Fortunately, the error message is reasonably clear. But I'd be happy to take a patch if you'd like to help improve matters.

**Comment by wiedemannc** [comment]
I've run into the same issue. This is preventing us from adding pycuda to a requirements.txt file which in turn is installed via pip -r requirements.txt.

**Comment by inducer** [comment]
I'd be happy to receive any suggestion on what I can change in PyCUDA in order to fix this. The point is that `import numpy` currently needs to succeed in the setup script in order to find the include path for numpy.

> This is preventing us from adding pycuda to a requirements.txt file which in turn is installed via pip -r requirements.txt.

Adding numpy above pycuda doesn't help?

**Comment by wiedemannc** [comment]
Unfortunately it doesn't. However, in the meantime I realized, that there are more packages suffering from similar issues. It seems that https://github.com/pypa/pip/issues/25 is discussing the issue and they seem to suggest to import build-time dependencies as late as possible in setup.py.

I've created a workaround now, we are processing the requirements.txt file line by line (such that we can control the order of the packages if needed). This seems to work out fine by now.

**Comment by inducer** [comment]
I think I tried that, and it didn't help.

**Comment by estan** [comment]
@inducer Ah yes, shame on me for not looking. It's even in `setup_requires` in `master`.

**Comment by estan** [comment]
@inducer I'm no distutils/setuptools expert, the one thing I can think of that would maybe solve this is to subclass the `build_ext` command, and do something like

```python
class custom_build_ext(build_ext):
    ...
    def build_extension(self, extension):
        ...
        if isinstance(extension, NumpyExtension):
            # Determine numpy flags here and add to extension
        ...
        super().build_extensions()
```

and then `cmdclass={'build_ext': custom_build_ext}` in the `setup` call to use the custom `build_ext`.

It's obviously quite hacky though, so not sure what you'd think of such a "solution"?

**Comment by estan** [comment]
@inducer A more concrete suggestion:

```diff
diff --git a/setup.py b/setup.py
index 775141b..97d2a4e 100644
--- a/setup.py
+++ b/setup.py
@@ -101,9 +101,29 @@ def get_config_schema():
 def main():
     import sys
 
+    from setuptools import Extension
+    from setuptools.command.build_ext import build_ext
     from aksetup_helper import (hack_distutils, get_config, setup,
-            NumpyExtension, set_up_shipped_boost_if_requested,
-            check_git_submodules)
+            set_up_shipped_boost_if_requested, check_git_submodules)
+
+    class NumpyExtension(Extension):
+        pass
+
+    class custom_build_ext(build_ext):
+
+        def build_extension(self, extension):
+            if isinstance(extension, NumpyExtension):
+                numpy_incpath = self.get_numpy_incpath()
+                if numpy_incpath not in extension.include_dirs:
+                    extension.include_dirs.append(numpy_incpath)
+            super().build_extension(extension)
+
+        def get_numpy_incpath(self):
+            from imp import find_module
+            # avoid actually importing numpy, it screws up distutils
+            file, pathname, descr = find_module("numpy")
+            from os.path import join
+            return join(pathname, "core", "include")
 
     check_git_submodules()
 
@@ -236,7 +256,7 @@ def main():
                     extra_link_args=conf["LDFLAGS"],
                     ),
                 ],
-
+            cmdclass={"build_ext": custom_build_ext},
             include_package_data=True,
             package_data={
                     "pycuda": [
```

With this I can pip install pycuda in an env without numpy.

**Comment by inducer** [comment]
I've done something I claim is equivalent to your PR in a6551dd442ed626d394d3431daa2fb1d1d3f3184. Please let me know if this works for you.

**Comment by estan** [comment]
@inducer Yup, works great:

```
(pycudaenv) [estan@newton pycuda (master=)]$ pip list
Package       Version
------------- -------
pip           18.1   
pkg-resources 0.0.0  
setuptools    40.6.3 
wheel         0.32.3 
(pycudaenv) [estan@newton pycuda (master=)]$ pip install -e .
Obtaining file:///home/estan/tmp/pycuda
Collecting pytools>=2011.2 (from pycuda==2018.1.1)
Collecting pytest>=2 (from pycuda==2018.1.1)
  Using cached https://files.pythonhosted.org/packages/19/80/1ac71d332302a89e8637456062186bf397abc5a5b663c1919b73f4d68b1b/pytest-4.0.2-py2.py3-none-any.whl
Collecting decorator>=3.2.0 (from pycuda==2018.1.1)
  Using cached https://files.pythonhosted.org/packages/bc/bb/a24838832ba35baf52f32ab1a49b906b5f82fb7c76b2f6a7e35e140bac30/decorator-4.3.0-py2.py3-none-any.whl
Collecting appdirs>=1.4.0 (from pycuda==2018.1.1)
  Using cached https://files.pythonhosted.org/packages/56/eb/810e700ed1349edde4cbdc1b2a21e28cdf115f9faf263f6bbf8447c1abf3/appdirs-1.4.3-py2.py3-none-any.whl
Collecting mako (from pycuda==2018.1.1)
Collecting numpy>=1.6.0 (from pytools>=2011.2->pycuda==2018.1.1)
  Using cached https://files.pythonhosted.org/packages/ff/7f/9d804d2348471c67a7d8b5f84f9bc59fd1cefa148986f2b74552f8573555/numpy-1.15.4-cp36-cp36m-manylinux1_x86_64.whl
Collecting six>=1.8.0 (from pytools>=2011.2->pycuda==2018.1.1)
  Using cached https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl
Collecting more-itertools>=4.0.0 (from pytest>=2->pycuda==2018.1.1)
  Using cached https://files.pythonhosted.org/packages/a4/a6/42f17d065bda1fac255db13afc94c93dbfb64393eae37c749b4cb0752fc7/more_itertools-5.0.0-py3-none-any.whl
Collecting pluggy>=0.7 (from pytest>=2->pycuda==2018.1.1)
  Using cached https://files.pythonhosted.org/packages/1c/e7/017c262070af41fe251401cb0d0e1b7c38f656da634cd0c15604f1f30864/pluggy-0.8.0-py2.py3-none-any.whl
Collecting py>=1.5.0 (from pytest>=2->pycuda==2018.1.1)
  Using cached https://files.pythonhosted.org/packages/3e/c7/3da685ef117d42ac8d71af525208759742dd235f8094221fdaafcd3dba8f/py-1.7.0-py2.py3-none-any.whl
Collecting atomicwrites>=1.0 (from pytest>=2->pycuda==2018.1.1)
  Using cached https://files.pythonhosted.org/packages/3a/9a/9d878f8d885706e2530402de6417141129a943802c084238914fa6798d97/atomicwrites-1.2.1-py2.py3-none-any.whl
Collecting attrs>=17.4.0 (from pytest>=2->pycuda==2018.1.1)
  Using cached https://files.pythonhosted.org/packages/3a/e1/5f9023cc983f1a628a8c2fd051ad19e76ff7b142a0faf329336f9a62a514/attrs-18.2.0-py2.py3-none-any.whl
Requirement already satisfied: setuptools in /home/estan/pycudaenv/lib/python3.6/site-packages (from pytest>=2->pycuda==2018.1.1) (40.6.3)
Collecting MarkupSafe>=0.9.2 (from mako->pycuda==2018.1.1)
  Using cached https://files.pythonhosted.org/packages/08/04/f2191b50fb7f0712f03f064b71d8b4605190f2178ba02e975a87f7b89a0d/MarkupSafe-1.1.0-cp36-cp36m-manylinux1_x86_64.whl
Installing collected packages: numpy, six, appdirs, decorator, pytools, more-itertools, pluggy, py, atomicwrites, attrs, pytest, MarkupSafe, mako, pycuda
  Running setup.py develop for pycuda
Successfully installed MarkupSafe-1.1.0 appdirs-1.4.3 atomicwrites-1.2.1 attrs-18.2.0 decorator-4.3.0 mako-1.0.7 more-itertools-5.0.0 numpy-1.15.4 pluggy-0.8.0 py-1.7.0 pycuda pytest-4.0.2 pytools-2018.5.2 six-1.12.0
(pycudaenv) [estan@newton pycuda (master=)]$
```

Thanks!

**Comment by estan** [comment]
@inducer Hey again, curious when we can enjoy this fix in a release on PyPI :) Have any idea when the next release might be made?

**Comment by inducer** [comment]
Sorry for the delay, and thanks for the reminder. I've just released 2019.1, which should include the fix.

**Comment by estan** [comment]
@inducer Great, no problem, much a appreciated.
